torch/csrc/jit/pybind_utils.h

```diff
@@ -88,82 +97,146 @@ inline c10::optional<TypePtr> unifyOrInitializeType(
   return unifyTypes(accum, unify);
 }
 
-inline TypedIValue toTypedIValue(py::handle input) {
+MatchTypeReturn tryToInferContainerType(py::handle input);
+
+// Try to infer the type of a Python object
+// The type cannot be inferred if:
+//   input is a None
+//   input is an empty container (list, dict, tuple)
+//   input is an empty container
+//   input is an dict with more than one type of key or value
+//   input is an list with more than one type of element
+inline MatchTypeReturn tryToInferType(py::handle input) {
+  // Try tensor types
   if (THPVariable_Check(input.ptr())) {
-    auto ten = py::cast<at::Tensor>(input);
-    if (ten.is_sparse()) {
-      AT_ERROR("sparse tensors not supported");
+    auto tensor = py::cast<at::Tensor>(input);
+    if (tensor.is_sparse()) {
+      return MatchTypeReturn("Sparse tensors not supported");
+    }
+    if (tensor.is_mkldnn()) {
+      // mkldnn tensor as opaque tensor doesn't have strides, so we can
+      // not create a CompleteTensorType
+      return MatchTypeReturn(DimensionedTensorType::create(tensor));
+    }
+    return MatchTypeReturn(CompleteTensorType::create(tensor));
+  }
+
+  if (input.is(py::none())) {
+    return MatchTypeReturn("Cannot infer type of a None value");
+  }
+
+  // Try basic types first
+  if (py::isinstance<py::bool_>(input)) {
+    return MatchTypeReturn(BoolType::get());
+  } else if (py::isinstance<py::int_>(input)) {
+    return MatchTypeReturn(IntType::get());
+  } else if (py::isinstance<py::float_>(input)) {
+    return MatchTypeReturn(FloatType::get());
+  } else if (py::isinstance<py::str>(input)) {
+    return MatchTypeReturn(StringType::get());
+  } else if (THPLayout_Check(input.ptr())) {
+    return MatchTypeReturn(IntType::get());
+  } else if (THPDevice_Check(input.ptr())) {
+    return MatchTypeReturn(DeviceObjType::get());
+  } else if (THPDtype_Check(input.ptr())) {
+    return MatchTypeReturn(IntType::get());
+  }
+
+  // Try container types
+  return tryToInferContainerType(input);
+}
+
+inline MatchTypeReturn tryToInferContainerType(py::handle input) {
+  if (THPVariable_Check(input.ptr())) {
```

Comment:
The tensor check is being duplicated, it is also in `tryToInferType`. Not sure why it is is in container inference